styx-common/src/test/java/com/spotify/styx/util/ShardedCounterTest.java

```diff
@@ -0,0 +1,299 @@
+                                                             .eq(PROPERTY_SHARD_VALUE,
+                                                                 1)))
+                                                     .build());
+    // assert there's one and only one shard with the value set to 1
+    assertEquals(1L, results.next().getLong(PROPERTY_SHARD_VALUE));
+    assertFalse(results.hasNext());
+
+    // assert the correct value is fetched after cache expiry
+    now = afterCacheExpiryDuration(now);
+    assertEquals(1L, shardedCounter.getCounter(COUNTER_ID1));
+  }
+
+  private void updateCounterInTransaction(String counterId1, long delta) {
+    datastore.runInTransaction(rw -> {
+      shardedCounter.updateCounter(rw, COUNTER_ID1, delta);
+      return null;
+    });
+  }
+
+  @Test
+  public void shoudDecrementCounter() {
+    now = Instant.parse("2018-01-01T00:00:00.000Z");
+    // init counter
+    assertEquals(0, shardedCounter.getCounter(COUNTER_ID1));
+
+    final int shardIndex = 3;
+    final Key shardKey = datastore.newKeyFactory().setKind(KIND_COUNTER_SHARD)
+        .newKey(COUNTER_ID1 + "-" + shardIndex);
+    //increment counter shard by 1
+    datastore.put(Entity.newBuilder(shardKey)
+                      .set(PROPERTY_COUNTER_ID, COUNTER_ID1)
+                      .set(PROPERTY_SHARD_INDEX, shardIndex)
+                      .set(PROPERTY_SHARD_VALUE, 1)
+                      .build());
+
+    now = afterCacheExpiryDuration(now);
+    // assert cache is updated with the new value
+    assertEquals(1L, shardedCounter.getCounter(COUNTER_ID1));
+
+    //decrement counter by 1
+    updateCounterInTransaction(COUNTER_ID1, -1L);
+
+    // assert that the only eligible shard was chosen to be decremented
+    assertEquals(0L, datastore.get(shardKey).getLong(PROPERTY_SHARD_VALUE));
+
+    // assert cache is updated with the new value
+    now = afterCacheExpiryDuration(now);
+    assertEquals(0L, shardedCounter.getCounter(COUNTER_ID1));
+  }
+
+
+  @Test(expected = DatastoreException.class)
+  public void shoudFailDecrementingEmptyCounter() {
+    now = Instant.parse("2018-01-01T00:00:00.000Z");
+    //increment counter by 1
+    updateCounterInTransaction(COUNTER_ID1, -1L);
+  }
+
+  @Test
+  public void shoudFailIncrementingFullCounter() {
+    now = Instant.parse("2018-01-01T00:00:00.000Z");
+    assertEquals(0L, shardedCounter.getCounter(COUNTER_ID1));
+    datastore.runInTransaction(transaction -> {
+      shardedCounter.updateLimit(transaction, COUNTER_ID1, 10);
+      return null;
+    });
+
+    //expire cache so that the new limit value gets picked up
+    now = afterCacheExpiryDuration(now);
+
+    //increment counter by 1 until counter value gets to 10
+    while (shardedCounter.getCounter(COUNTER_ID1) < 10) {
+      try {
+        IntStream.range(0, 10).forEach(i -> updateCounterInTransaction(COUNTER_ID1, 1L));
+      } catch (DatastoreException ignored) {
```

Comment:
What were these exceptions about?

Reply:
This is made to catch exceptions caused by collisions, or bad shard selection, or simply exceptions caused by no capacity left in a counter. The goal is to fill up all capacity of the counter (in this case capacity limit is set to 10). It runs in loops to make sure the we get up to the max capacity of the counter, and not fail early due to selecting a full shard.

Reply:
Ah I misread the line above as doing _one_ increment.

If it did so then, then exceptions would not occur, I think? (We bump `now` in each iteration, and 100% consistency is used in `helper`, and no other tests run in parallel, so in that setup collisions and bad shard selection, and exceeded capacity are not expected to happen by design.)